qiskit/transpiler/preset_passmanagers/level2.py

```diff
@@ -227,16 +230,34 @@ def _opt_control(property_set):
         pre_optimization = common.generate_pre_op_passmanager(target, coupling_map, True)
     else:
         pre_optimization = common.generate_pre_op_passmanager(remove_reset_in_zero=True)
-    optimization = PassManager()
-    unroll = [pass_ for x in translation.passes() for pass_ in x["passes"]]
-    optimization.append(_depth_check + _size_check)
-    opt_loop = _opt + unroll + _depth_check + _size_check
-    optimization.append(opt_loop, do_while=_opt_control)
-    sched = common.generate_scheduling(
-        instruction_durations, scheduling_method, timing_constraints, inst_map
-    )
+    if optimization_method is None:
+        optimization = PassManager()
+        unroll = [pass_ for x in translation.passes() for pass_ in x["passes"]]
+        optimization.append(_depth_check + _size_check)
+        opt_loop = _opt + unroll + _depth_check + _size_check
+        optimization.append(opt_loop, do_while=_opt_control)
+    else:
+        optimization = plugin_manager.get_passmanager_stage(
+            "optimization", optimization_method, pass_manager_config
+        )
+    if scheduling_method is None or scheduling_method in {"alap", "asap"}:
+        sched = common.generate_scheduling(
+            instruction_durations, scheduling_method, timing_constraints, inst_map
+        )
+    else:
+        sched = plugin_manager.get_passmanager_stage(
+            "scheduling", scheduling_method, pass_manager_config
+        )
+    if init_method is not None:
+        init = plugin_manager.get_passmanager_stage("init", init_method, pass_manager_config)
+    else:
+        init = unroll_3q
+
+    # Restore PassManagerConfig optimization_level override
+    pass_manager_config.optimization_level = optimization_level
```

Comment:
I was doing this to avoid potentially mutating the pass manager config as a side effect of the pass manager generation in these function. Since the `optimization_level` field is an optional new addition to a passmanager config it has been totally valid to do something like:

```
pm_config = PassManagerConfig.from_backend(backend)
print(pm_config)
pm_level2 = level_2_pass_manager(pm_config)
print(pm_config)
```
without an error. If I were running that I'd have expected `pm_config` to remain unchanged. But now with the plugins the plugin potentially needs the optimization level context to know how to build the stage correctly which we get implicitly from the level 2 function.

This code only has an effect if the input value is `None` because we temporarily changed the internal value for the plugins and restore it to the original. If you did something like:

```
pm_config = PassManagerConfig.from_backend(backend, optimization_level=3)
print(pm_config.optimization_level)
pm_level2 = level_2_pass_manager(pm_config)
print(pm_config.optimization_level)
```
it'd print 3 both times because we're basically doing `3 = 3` here.